metaflow/plugins/aws/awslambda/lambda_runner.py

```diff
@@ -0,0 +1,293 @@
+            self._echo("[ temporary error in fetching logs: %s ]" % ex, True)
+
+    def run(self):
+        delayed_update_while(
+            condition=lambda: not self._stop_flag,
+            update_fn=self._print_available_from_s3,
+        )
+        self._print_available_from_s3()
+
+
+def lambda_name(memory_mb, timeout_seconds, image_uri):
+    """ Construct lambda name based on hashed resources """
+    res_hash = hashlib.sha256(
+        str(memory_mb).encode("utf8")
+        + str(timeout_seconds).encode("utf8")
+        + image_uri.encode("utf8")
+    ).hexdigest()[:10]
+
+    # We append unhashed resource values here for a bit more debuggability
+    return "-".join(["mflambda", res_hash, str(memory_mb), str(timeout_seconds)])
+
+
+def ensure_lambda(
+    name,  # type: str
+    timeout_seconds,  # type: int
+    memory_mb,  # type: int
+    role_arn,
+    image_uri,
+    refresh_image,  # type: bool
+    echo,  # type: Callable[[str, bool], None]
+):
+    # type: (...) -> str
+    """
+    Make sure Lambda for the step exists. If it does, but resources or timeout
+    configuration is not what we expect, update the lambda configuration.
+
+    Return lambda ARN.
+    """
+
+    client = get_aws_client("lambda")
+
+    # Since multiple tasks in a foreach branch reuse the same Lambda, they all might
+    # get to this point simultaneously and try to update lambda configuration at the
+    # same time. For simplicity, we don't do any sychronization between them, and simply
+    # retry on conflict, as these updates should be idempotent anyway.
+    num_retries_on_conflict = 6
+
+    for retry in range(num_retries_on_conflict):
+        try:
+            return _maybe_update_lambda(
+                client,
+                name,
+                timeout_seconds,
+                memory_mb,
+                role_arn,
+                image_uri,
+                refresh_image,
+                echo=echo,
+            )
+        except client.exceptions.ResourceConflictException:
+            if retry == num_retries_on_conflict - 1:
+                raise
+            else:
+                waiter = client.get_waiter("function_updated")
+                waiter.wait(FunctionName=name)
+    else:
+        # shouldn't really get here ever
+        raise Exception("Failed to update the function")
+
+
+def _maybe_update_lambda(
+    client,
+    name,
+    timeout_seconds,  # type: int
+    memory_size,  # type: float
+    role_arn,
+    image_uri,
+    refresh_image,  # type: bool
+    echo,  # type: Callable[[str, bool], None]
+):
+    # type: (...) -> str
+    """
+    This does all the necessary AWS API Calls to update or create Lambda. Returns
+    Lambda ARN.
+    """
+    try:
+        # Get current function configuration
+        info = client.get_function(FunctionName=name)
+
+        # If refresh_image is set, force update function code.
+        if refresh_image:
+            echo("Updating image for %s.." % name, False)
```

Comment:
Given that this is an internal implementation detail - should we print this? Same comment for L:178.

Reply:
I kinda went back and forth on this; the main reason for having this is that it can often take about 30 seconds to update or create lambda (if it doesn't exist or `refresh_image` is used). So I wanted to give the user some indication that something is happening, Metaflow isn't magically "stuck".

Reply:
We can have a different message to address that use case. For example - AWS Batch cycles through the batch status codes so that the user understands that the job is in progress - we print a message when the status changes or if it's been 30s since the last message. We can have a similar behavior here.

Reply:
Can we have the same(similar) message as we have for AWS Batch?

Reply:
Hmm the batch is a different beast so the message is different, there is no "job status" in the case of Lambda. And note that is not a direct equivalent of what Batch is doing, you'll only see this message on the first run when Lambda resource needs to be created.

So the copy here is the best I could come up with -- but again, open to concrete alternative suggestions